File: python/bench/bench_pipeline.py
#!/usr/bin/env python3
"""In-process latency benchmark for the Python pipeline.

Same command line as `cpp/build/bench_pipeline`, same two output files, so
`harness/compare_outputs.py` and the analysis do not need to know which
implementation produced which.

Three things this runner does deliberately:

* The frame store is read into RAM before the first measurement. A benchmark
  that reads a file per frame measures the page cache.
* Records are formatted and written after the loop, never inside it, so no
  measured frame pays for `json.dumps` or a `write`.
* Garbage collection is counted, not disabled. A real rclpy node runs with the
  collector on, so the honest thing is to leave it on and record when it fires.
  The counter is a `gc.callbacks` hook rather than `gc.get_stats()`, because
  `get_stats` builds a list of dicts on every call and the harness must not be
  a meaningful contributor to the allocation pressure it is trying to measure.

The measured frames cycle through the store, so the working set is the store
rather than one frame sitting hot in L2.
"""
from __future__ import annotations

import argparse
import gc
import hashlib
import json
import os
import platform
import sys
import time
from pathlib import Path

import numpy as np

REPO_ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO_ROOT / 'python'))

from grasp_core import (STAGE_KEYS, GraspPipeline,  # noqa: E402
                        calibrate_timer_ns)

SIGNIFICANT = '.17g'
CHECKSUM_DECIMALS = 9
PERCENTILES = (50.0, 95.0, 99.0)
# Percentiles are reported at the nearest observed sample rather than
# interpolated between two: a latency that was never measured is not a latency.
PERCENTILE_METHOD = 'nearest'


def number(value) -> str:
    """A double that survives a JSON round trip, per docs/FORMATS.md."""
    return format(float(value), SIGNIFICANT)


def vector(values) -> str:
    return '[' + ','.join(number(v) for v in values) + ']'


def flag(value) -> str:
    return 'true' if value else 'false'


class FrameStore:
    """The whole corpus in memory, as the bytes a subscriber would receive."""

    def __init__(self, directory: Path):
        manifest = json.loads((directory / 'manifest.json').read_text())
        self.name = manifest['name']
        self.width = manifest['width']
        self.height = manifest['height']
        self.manifest = manifest
        self.frames = [
            ((directory / entry['depth']).read_bytes(),
             (directory / entry['rgb']).read_bytes())
            for entry in manifest['frames']]
        expected = self.width * self.height * 2
        for depth, _ in self.frames:
            if len(depth) != expected:
                raise ValueError(f'{directory} holds a depth frame of '
                                 f'{len(depth)} bytes, expected {expected}')

    def __len__(self) -> int:
        return len(self.frames)


class CollectionCounter:
    """Cumulative CPython collections per generation, per docs/FORMATS.md.

    A `gc.callbacks` hook costs one Python call per collection, where sampling
    `gc.get_stats()` would cost a list and three dicts per frame.
    """

    def __init__(self):
        self.counts = [0, 0, 0]
        gc.callbacks.append(self._on_collect)

    def _on_collect(self, phase, info):
        if phase == 'stop':
            self.counts[info['generation']] += 1

    def detach(self):
        gc.callbacks.remove(self._on_collect)


def trajectory_checksum(result) -> str:
    """SHA-256 over the waypoint block, exactly as docs/FORMATS.md defines it.

    Positions, then velocities, then accelerations, then time, waypoint by
    waypoint, each an IEEE-754 double rounded to nine decimals first, because
    the two implementations only agree to 1e-6.
    """
    count = result.positions.shape[0]
    block = np.empty((count, result.positions.shape[1] * 3 + 1),
                     dtype=np.float64)
    dof = result.positions.shape[1]
    block[:, :dof] = result.positions
    block[:, dof:2 * dof] = result.velocities
    block[:, 2 * dof:3 * dof] = result.accelerations
    block[:, 3 * dof] = result.times
    np.round(block, CHECKSUM_DECIMALS, out=block)
    return hashlib.sha256(block.tobytes()).hexdigest()


def output_record(impl: str, frame: int, result) -> str:
    return (
        '{"impl":"' + impl + '","frame":' + str(frame)
        + ',"plane_found":' + flag(result.plane_found)
        + ',"plane":' + vector(result.plane)
        + ',"graspable":' + flag(result.graspable)
        + ',"width":' + number(result.width)
        + ',"tcp":' + vector(result.tcp.reshape(-1))
        + ',"converged":' + flag(result.converged)
        + ',"iterations":' + str(result.iterations)
        + ',"q":' + vector(result.q)
        + ',"duration_s":' + number(result.duration_s)
        + ',"traj_checksum":"' + trajectory_checksum(result) + '"}')


def timing_record(impl: str, dataset: str, row) -> str:
    (seq, frame, total, points, cluster_points, iterations,
     plane_found, graspable, converged, gen0, gen1, gen2) = row[:12]
    stages = row[12:]
    parts = ','.join(f'"{key}":{value}' for key, value in zip(STAGE_KEYS, stages))
    return (
        '{"impl":"' + impl + '","dataset":"' + dataset + '","frame":'
        + str(frame) + ',"seq":' + str(seq)
        + ',"stage_ns":{' + parts + '}'
        + ',"total_ns":' + str(total)
        + ',"points":' + str(points)
        + ',"cluster_points":' + str(cluster_points)
        + ',"ik_iterations":' + str(iterations)
        + ',"plane_found":' + flag(plane_found)
        + ',"graspable":' + flag(graspable)
        + ',"converged":' + flag(converged)
        + ',"gc":{"gen0":' + str(gen0) + ',"gen1":' + str(gen1)
        + ',"gen2":' + str(gen2) + '}}')


def percentiles(samples: np.ndarray) -> dict:
    return {f'p{int(p)}_ns': float(np.percentile(samples, p,
                                                 method=PERCENTILE_METHOD))
            for p in PERCENTILES}


def main(argv=None) -> int:
    default_config = REPO_ROOT / 'assets' / 'pipeline_config.json'
    config = json.loads(default_config.read_text())
    benchmark = config['benchmark']

    parser = argparse.ArgumentParser(description=__doc__,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument('--dataset', type=Path, required=True,
                        help='frame store directory, holding manifest.json')
    parser.add_argument('--out-timing', type=Path, required=True)
    parser.add_argument('--out-output', type=Path, required=True)
    parser.add_argument('--out-meta', type=Path, default=None,
                        help='run metadata; defaults beside --out-timing')
    parser.add_argument('--warmup', type=int, default=benchmark['warmup_frames'])
    parser.add_argument('--frames', type=int, default=benchmark['measured_frames'])
    parser.add_argument('--config', type=Path, default=default_config)
    parser.add_argument('--chain', type=Path,
                        default=REPO_ROOT / 'assets' / 'franka' / 'panda_chain.json')
    parser.add_argument('--ransac-table', type=Path,
                        default=REPO_ROOT / 'assets' / 'ransac_uniform.bin')
    parser.add_argument('--impl', default='py')
    parser.add_argument('--quiet', action='store_true')
    args = parser.parse_args(argv)

    store = FrameStore(args.dataset)
    pipeline = GraspPipeline(str(args.config), str(args.chain),
                             str(args.ransac_table))
    timer = calibrate_timer_ns()

    frames = args.frames
    stages = np.zeros((frames, len(STAGE_KEYS)), dtype=np.int64)
    scalars = np.zeros((frames, 6), dtype=np.int64)
    flags = np.zeros((frames, 3), dtype=np.bool_)
    collections = np.zeros((frames, 3), dtype=np.int64)
    outputs = {}

    counter = CollectionCounter()
    counts = counter.counts
    baseline = None
    width = store.width
    height = store.height
    corpus = store.frames
    total_frames = len(corpus)
    run = pipeline.run
    impl = args.impl

    for index in range(args.warmup):
        depth, colour = corpus[index % total_frames]
        run(depth, colour, width, height)

    baseline = list(counts)
    started = time.time()
    for seq in range(frames):
        frame = (args.warmup + seq) % total_frames
        depth, colour = corpus[frame]
        result = run(depth, colour, width, height)

        stage_ns = result.stage_ns
        row = stages[seq]
        row[0] = stage_ns['decode']
        row[1] = stage_ns['deproject']
        row[2] = stage_ns['transform_crop']
        row[3] = stage_ns['plane']
        row[4] = stage_ns['cluster']
        row[5] = stage_ns['grasp']
        row[6] = stage_ns['ik']
        row[7] = stage_ns['traj']
        row = scalars[seq]
        row[0] = seq
        row[1] = frame
        row[2] = result.total_ns
        row[3] = result.points
        row[4] = result.cluster_points
        row[5] = result.iterations
        row = flags[seq]
        row[0] = result.plane_found
        row[1] = result.graspable
        row[2] = result.converged
        row = collections[seq]
        row[0] = counts[0]
        row[1] = counts[1]
        row[2] = counts[2]

        if frame not in outputs:
            outputs[frame] = output_record(impl, frame, result)
    elapsed = time.time() - started
    counter.detach()

    args.out_timing.parent.mkdir(parents=True, exist_ok=True)
    with args.out_timing.open('w', encoding='utf-8') as handle:
        for seq in range(frames):
            row = (list(scalars[seq]) + list(flags[seq])
                   + list(collections[seq]) + list(stages[seq]))
            handle.write(timing_record(impl, store.name, row) + '\n')

    args.out_output.parent.mkdir(parents=True, exist_ok=True)
    with args.out_output.open('w', encoding='utf-8') as handle:
        for frame in sorted(outputs):
            handle.write(outputs[frame] + '\n')

    totals = scalars[:, 2]
    summary = {
        'impl': impl,
        'dataset': store.name,
        'resolution': [width, height],
        'store_frames': total_frames,
        'warmup_frames': args.warmup,
        'measured_frames': frames,
        'wall_clock_s': elapsed,
        'percentile_method': PERCENTILE_METHOD,
        'total_ns': percentiles(totals),
        'stage_ns': {key: percentiles(stages[:, i])
                     for i, key in enumerate(STAGE_KEYS)},
        'gc_collections_total': [int(counts[i] - baseline[i]) for i in range(3)],
        'gc_collections_per_100_frames': [
            (counts[i] - baseline[i]) * 100.0 / frames for i in range(3)],
        'bail_outs': {
            'plane_not_found': int((~flags[:, 0]).sum()),
            'not_graspable': int((~flags[:, 1]).sum()),
            'ik_not_converged': int((~flags[:, 2]).sum()),
        },
        'ik_iterations': {
            'median': float(np.median(scalars[:, 5])),
            'max': int(scalars[:, 5].max()),
        },
        'points': {'median': float(np.median(scalars[:, 3]))},
        'timer': timer,
        'environment': {
            'python': platform.python_version(),
            'numpy': np.__version__,
            'platform': platform.platform(),
            'processor': platform.processor(),
            'cpu_count': os.cpu_count(),
            'load_average': list(os.getloadavg()),
            'blas_threads': {name: os.environ.get(name)
                             for name in ('OPENBLAS_NUM_THREADS',
                                          'OMP_NUM_THREADS',
                                          'MKL_NUM_THREADS')},
        },
        'inputs': {
            'config': str(args.config),
            'chain': str(args.chain),
            'ransac_table': str(args.ransac_table),
        },
    }
    meta_path = args.out_meta
    if meta_path is None:
        name = args.out_timing.name
        stem = name[:-len('.timing.jsonl')] if name.endswith('.timing.jsonl') \
            else args.out_timing.stem
        meta_path = args.out_timing.with_name(stem + '.meta.json')
    meta_path.write_text(json.dumps(summary, indent=2) + '\n', encoding='utf-8')

    if not args.quiet:
        report(summary, stages, totals)
    return 0


def report(summary, stages, totals) -> None:
    print(f"{summary['impl']} on {summary['dataset']} "
          f"{summary['resolution'][0]}x{summary['resolution'][1]}, "
          f"{summary['measured_frames']} measured frames "
          f"in {summary['wall_clock_s']:.1f} s")
    print(f"  {'stage':<16}{'p50 ms':>10}{'p95 ms':>10}{'p99 ms':>10}"
          f"{'share':>8}")
    median_total = summary['total_ns']['p50_ns']
    for i, key in enumerate(STAGE_KEYS):
        stage = summary['stage_ns'][key]
        share = 100.0 * stage['p50_ns'] / median_total if median_total else 0.0
        print(f"  {key:<16}{stage['p50_ns'] / 1e6:>10.3f}"
              f"{stage['p95_ns'] / 1e6:>10.3f}{stage['p99_ns'] / 1e6:>10.3f}"
              f"{share:>7.1f}%")
    print(f"  {'total':<16}{median_total / 1e6:>10.3f}"
          f"{summary['total_ns']['p95_ns'] / 1e6:>10.3f}"
          f"{summary['total_ns']['p99_ns'] / 1e6:>10.3f}")
    print(f"  timer read: median {summary['timer']['median_ns']:.0f} ns "
          f"over {summary['timer']['samples']} samples")
    print(f"  gc collections per 100 frames: "
          f"{summary['gc_collections_per_100_frames']}")
    print(f"  bail-outs: {summary['bail_outs']}, "
          f"ik iterations median {summary['ik_iterations']['median']:.0f} "
          f"max {summary['ik_iterations']['max']}")


if __name__ == '__main__':
    sys.exit(main())

File: cpp/bench/bench_pipeline.cpp
// In-process benchmark for the C++ core. The whole frame store is read into
// memory before the loop, both output files are written after it, and nothing
// between the two allocates, formats a string or touches a file: the point of
// the exercise is to time the pipeline, not the harness around it.
//
//   bench_pipeline --dataset data/table_640x480
//                  --out-timing results/cpp.timing.jsonl
//                  --out-output results/cpp.output.jsonl
//                  --warmup 100 --frames 2000

#include <array>
#include <cinttypes>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <stdexcept>
#include <iostream>
#include <string>
#include <vector>

#include "grasp_core/config.hpp"
#include "grasp_core/json.hpp"
#include "grasp_core/pipeline.hpp"

namespace
{

struct Options
{
  std::string dataset;
  std::string out_timing;
  std::string out_output;
  std::string config_path{"assets/pipeline_config.json"};
  std::string chain_path{"assets/franka/panda_chain.json"};
  std::string ransac_path{"assets/ransac_uniform.bin"};
  std::string impl{"cpp"};
  int warmup{-1};
  int frames{-1};
};

[[noreturn]] void usage(const char * program, const std::string & why)
{
  std::cerr << why << "\n"
            << "usage: " << program << " --dataset DIR --out-timing FILE --out-output FILE\n"
            << "       [--warmup N] [--frames N] [--impl NAME]\n"
            << "       [--config FILE] [--chain FILE] [--ransac-table FILE]\n";
  std::exit(2);
}

Options parse_options(int argc, char ** argv)
{
  Options o;
  for (int i = 1; i < argc; ++i) {
    const std::string flag = argv[i];
    const bool has_value = (i + 1) < argc;
    auto value = [&]() -> std::string {
        if (!has_value) {
          usage(argv[0], "missing value for " + flag);
        }
        return argv[++i];
      };
    if (flag == "--dataset") {
      o.dataset = value();
    } else if (flag == "--out-timing") {
      o.out_timing = value();
    } else if (flag == "--out-output") {
      o.out_output = value();
    } else if (flag == "--config") {
      o.config_path = value();
    } else if (flag == "--chain") {
      o.chain_path = value();
    } else if (flag == "--ransac-table") {
      o.ransac_path = value();
    } else if (flag == "--impl") {
      o.impl = value();
    } else if (flag == "--warmup") {
      o.warmup = std::atoi(value().c_str());
    } else if (flag == "--frames") {
      o.frames = std::atoi(value().c_str());
    } else {
      usage(argv[0], "unknown flag " + flag);
    }
  }
  if (o.dataset.empty() || o.out_timing.empty() || o.out_output.empty()) {
    usage(argv[0], "--dataset, --out-timing and --out-output are all required");
  }
  return o;
}

struct FrameStore
{
  std::string name;
  int width{0};
  int height{0};
  int count{0};
  std::vector<std::uint16_t> depth;   // count * width * height
  std::vector<std::uint8_t> rgb;      // count * width * height * 3
};

void read_exact(const std::string & path, void * dst, std::size_t bytes)
{
  std::ifstream in(path, std::ios::binary);
  if (!in) {
    throw std::runtime_error("cannot open " + path);
  }
  in.read(static_cast<char *>(dst), static_cast<std::streamsize>(bytes));
  if (static_cast<std::size_t>(in.gcount()) != bytes) {
    throw std::runtime_error(path + " is shorter than the manifest says");
  }
}

FrameStore load_store(const std::string & directory)
{
  using grasp_core::json::Value;
  const Value manifest = grasp_core::json::parse_file(directory + "/manifest.json");

  FrameStore store;
  store.name = manifest["name"].string();
  store.width = manifest["width"].integer();
  store.height = manifest["height"].integer();
  const auto & frames = manifest["frames"].array();
  store.count = static_cast<int>(frames.size());
  if (store.count < 1) {
    throw std::runtime_error(directory + " lists no frames");
  }

  const std::size_t pixels = static_cast<std::size_t>(store.width) *
    static_cast<std::size_t>(store.height);
  store.depth.resize(pixels * static_cast<std::size_t>(store.count));
  store.rgb.resize(pixels * 3u * static_cast<std::size_t>(store.count));

  for (int i = 0; i < store.count; ++i) {
    const auto & entry = frames[static_cast<std::size_t>(i)];
    read_exact(
      directory + "/" + entry["depth"].string(),
      store.depth.data() + pixels * static_cast<std::size_t>(i),
      pixels * sizeof(std::uint16_t));
    read_exact(
      directory + "/" + entry["rgb"].string(),
      store.rgb.data() + pixels * 3u * static_cast<std::size_t>(i),
      pixels * 3u);
  }
  return store;
}

struct TimingRecord
{
  int frame{0};
  int seq{0};
  grasp_core::StageNs stage;
  std::uint64_t total_ns{0};
  int points{0};
  int cluster_points{0};
  int ik_iterations{0};
  bool plane_found{false};
  bool graspable{false};
  bool converged{false};
};

struct OutputRecord
{
  bool seen{false};
  bool plane_found{false};
  std::array<double, 4> plane{};
  int cluster_points{0};
  bool graspable{false};
  double width{0.0};
  std::array<double, 16> tcp{};
  bool converged{false};
  int iterations{0};
  std::array<double, grasp_core::kDof> q{};
  double duration_s{0.0};
  // waypoints * (3 * kDof + 1), in the order docs/FORMATS.md fixes: one
  // waypoint's positions, velocities, accelerations and time, then the next.
  std::vector<double> waypoint_block;
};

// %.17g so a double survives a round trip through JSON unchanged.
void append_double(std::string & out, double v)
{
  char buffer[48];
  std::snprintf(buffer, sizeof(buffer), "%.17g", v);
  out += buffer;
}

void append_doubles(std::string & out, const double * values, std::size_t n)
{
  out += '[';
  for (std::size_t i = 0; i < n; ++i) {
    if (i != 0) {
      out += ',';
    }
    append_double(out, values[i]);
  }
  out += ']';
}

void append_int(std::string & out, long long v)
{
  char buffer[32];
  std::snprintf(buffer, sizeof(buffer), "%lld", v);
  out += buffer;
}

void append_bool(std::string & out, bool v)
{
  out += v ? "true" : "false";
}

}  // namespace

int main(int argc, char ** argv)
{
  try {
    const Options opt = parse_options(argc, argv);
    const FrameStore store = load_store(opt.dataset);

    grasp_core::Pipeline pipeline(opt.config_path, opt.chain_path, opt.ransac_path);
    const grasp_core::Config & cfg = pipeline.config();

    const int warmup = opt.warmup >= 0 ? opt.warmup : cfg.benchmark.warmup_frames;
    const int measured = opt.frames > 0 ? opt.frames : cfg.benchmark.measured_frames;
    const std::size_t pixels = static_cast<std::size_t>(store.width) *
      static_cast<std::size_t>(store.height);
    const std::size_t block_size =
      static_cast<std::size_t>(cfg.trajectory.waypoints) * (3u * grasp_core::kDof + 1u);

    // Everything the loop will touch is allocated here, including the
    // pipeline's own scratch, so the first measured frame is no different from
    // the last one.
    pipeline.reserve(store.width, store.height);
    std::vector<TimingRecord> timings(static_cast<std::size_t>(measured));
    std::vector<OutputRecord> outputs(static_cast<std::size_t>(store.count));
    for (auto & record : outputs) {
      record.waypoint_block.resize(block_size);
    }

    const int total_frames = warmup + measured;
    for (int i = 0; i < total_frames; ++i) {
      const int frame = i % store.count;
      const grasp_core::Result & result = pipeline.run(
        store.depth.data() + pixels * static_cast<std::size_t>(frame),
        store.rgb.data() + pixels * 3u * static_cast<std::size_t>(frame),
        store.width, store.height);
      if (i < warmup) {
        continue;
      }

      const int seq = i - warmup;
      TimingRecord & t = timings[static_cast<std::size_t>(seq)];
      t.frame = frame;
      t.seq = seq;
      t.stage = result.stage_ns;
      t.total_ns = result.total_ns;
      t.points = result.points;
      t.cluster_points = result.cluster_points;
      t.ik_iterations = result.iterations;
      t.plane_found = result.plane_found;
      t.graspable = result.graspable;
      t.converged = result.converged;

      // The pipeline is deterministic, so a frame seen twice produces the same
      // outputs; recording it once keeps the equivalence file one line per
      // frame of the store rather than one line per measured sample.
      OutputRecord & o = outputs[static_cast<std::size_t>(frame)];
      if (!o.seen) {
        o.seen = true;
        o.plane_found = result.plane_found;
        o.plane = result.plane;
        o.cluster_points = result.cluster_points;
        o.graspable = result.graspable;
        o.width = result.width;
        o.tcp = result.tcp;
        o.converged = result.converged;
        o.iterations = result.iterations;
        o.q = result.q;
        o.duration_s = result.duration_s;
        std::size_t at = 0;
        for (const auto & w : result.trajectory) {
          for (int j = 0; j < grasp_core::kDof; ++j) {
            o.waypoint_block[at++] = w.position[static_cast<std::size_t>(j)];
          }
          for (int j = 0; j < grasp_core::kDof; ++j) {
            o.waypoint_block[at++] = w.velocity[static_cast<std::size_t>(j)];
          }
          for (int j = 0; j < grasp_core::kDof; ++j) {
            o.waypoint_block[at++] = w.acceleration[static_cast<std::size_t>(j)];
          }
          o.waypoint_block[at++] = w.time_from_start;
        }
      }
    }

    const long long overhead_ns = static_cast<long long>(std::llround(
        pipeline.timer_overhead_ns()));

    std::string text;
    text.reserve(static_cast<std::size_t>(measured) * 400u);
    for (const TimingRecord & t : timings) {
      text += "{\"impl\":\"" + opt.impl + "\",\"dataset\":\"" + store.name + "\",\"frame\":";
      append_int(text, t.frame);
      text += ",\"seq\":";
      append_int(text, t.seq);
      text += ",\"stage_ns\":{\"decode\":";
      append_int(text, static_cast<long long>(t.stage.decode));
      text += ",\"deproject\":";
      append_int(text, static_cast<long long>(t.stage.deproject));
      text += ",\"transform_crop\":";
      append_int(text, static_cast<long long>(t.stage.transform_crop));
      text += ",\"plane\":";
      append_int(text, static_cast<long long>(t.stage.plane));
      text += ",\"cluster\":";
      append_int(text, static_cast<long long>(t.stage.cluster));
      text += ",\"grasp\":";
      append_int(text, static_cast<long long>(t.stage.grasp));
      text += ",\"ik\":";
      append_int(text, static_cast<long long>(t.stage.ik));
      text += ",\"traj\":";
      append_int(text, static_cast<long long>(t.stage.traj));
      text += "},\"total_ns\":";
      append_int(text, static_cast<long long>(t.total_ns));
      text += ",\"points\":";
      append_int(text, t.points);
      text += ",\"cluster_points\":";
      append_int(text, t.cluster_points);
      text += ",\"ik_iterations\":";
      append_int(text, t.ik_iterations);
      text += ",\"plane_found\":";
      append_bool(text, t.plane_found);
      text += ",\"graspable\":";
      append_bool(text, t.graspable);
      text += ",\"converged\":";
      append_bool(text, t.converged);
      text += ",\"timer_overhead_ns\":";
      append_int(text, overhead_ns);
      text += "}\n";
    }
    {
      std::ofstream out(opt.out_timing, std::ios::binary);
      if (!out) {
        throw std::runtime_error("cannot write " + opt.out_timing);
      }
      out << text;
    }

    text.clear();
    // Roughly 24 characters per %.17g double, plus the fixed fields. One
    // reserve rather than a dozen reallocations while the block is appended.
    text.reserve(static_cast<std::size_t>(store.count) * (24u * (block_size + 32u)));
    for (int frame = 0; frame < store.count; ++frame) {
      const OutputRecord & o = outputs[static_cast<std::size_t>(frame)];
      if (!o.seen) {
        continue;
      }
      text += "{\"impl\":\"" + opt.impl + "\",\"frame\":";
      append_int(text, frame);
      text += ",\"plane_found\":";
      append_bool(text, o.plane_found);
      text += ",\"plane\":";
      append_doubles(text, o.plane.data(), o.plane.size());
      text += ",\"cluster_points\":";
      append_int(text, o.cluster_points);
      text += ",\"graspable\":";
      append_bool(text, o.graspable);
      text += ",\"width\":";
      append_double(text, o.width);
      text += ",\"tcp\":";
      append_doubles(text, o.tcp.data(), o.tcp.size());
      text += ",\"converged\":";
      append_bool(text, o.converged);
      text += ",\"iterations\":";
      append_int(text, o.iterations);
      text += ",\"q\":";
      append_doubles(text, o.q.data(), o.q.size());
      text += ",\"duration_s\":";
      append_double(text, o.duration_s);
      text += ",\"trajectory\":";
      append_doubles(text, o.waypoint_block.data(), o.waypoint_block.size());
      text += "}\n";
    }
    {
      std::ofstream out(opt.out_output, std::ios::binary);
      if (!out) {
        throw std::runtime_error("cannot write " + opt.out_output);
      }
      out << text;
    }

    std::cerr << "dataset=" << store.name << " " << store.width << "x" << store.height
              << " frames=" << store.count << " warmup=" << warmup
              << " measured=" << measured
              << " timer_overhead_ns=" << overhead_ns << "\n";
    return 0;
  } catch (const std::exception & e) {
    std::cerr << "bench_pipeline: " << e.what() << "\n";
    return 1;
  }
}
